torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -246,8 +246,9 @@ def init_all_gather_output(
         if self.all_gather_output.numel() > 0:
             return  # already initialized
         all_gather_output_size = torch.Size([all_gather_input_numel * world_size])
+        # NOTE(yf225): here we explicitly make fsdp_param.all_gather_output the leaf tensor (instead of unsharded_param being leaf tensor)
```

Comment:
I want to be careful about this for the long term solution.

The reason we have a distinction between `all_gather_output` and `unsharded_param` is because for certain extensions, they could be different. For example, if we were to support sharding QLoRA parameters, then there could be multiple `all_gather_output`s, but they will get transformed into a single subclass that is the `unsharded_param`.

However, for the default `torch.Tensor` case, `unsharded_param` should share the same data as `all_gather_output` but have any padding removed.

Reply:
Discussed with Andrew offline: making `fsdp_param.all_gather_output` (i.e. padded unsharded) the requires_grad=True leaf tensor and making `unsharded_param` (i.e. unpadded unsharded) a differentiable view into it is okay to do.